### PR TITLE
adding lots of new options!

### PR DESCRIPTION
**Pull request by ChangePlaces**
- Added `--pagination` option to the crud:generate and crud:controller options to change amount of models shown per page on index page, e.g. `crud:generate --fields="...." --pagination=50`

-  added `--indexes` option (see #102)
e.g. `crud:generate --fields="title#string,some_value#integer" --indexes="some_value"` which will result in a migration of `$table->index('some_value');`. You can also do `--indexes="some_value#unique"` which will result in `$table->unique('some_value');`.

- added `--required-fields` option to allow for required fields to be specified (this means it's now ignoring any #req or #required that may be defined in the fields string. This helps make the crud definition easier to write and understand. e.g. `crud:generate --fields="title#string,value#integer" --required-fields="title"`. Fields that are not required will be set to `nullable`

- added `--foreign-keys` option to migration and generate command. This will create references in the migration table only, and won't create any relationship functions in the model class - that will come next! e.g. `--foreign-keys="owner_id#id#owners"` where owner_id is the field name of the current table, id is the id column name on the related table, and owners is the related table name.

- added `--relationships` option to auto-create relationship methods. This will create the necessary functions in the model class for relationships. Use; `--relationships="relationshipName#relationshipType#argsSeparatedByPipes"` where relationshipType is hasOne, belongsTo etc. User is responsible for ensuring the relationship is valid. The args will be split by pipe, and fed into the function. e.g. `--relationships="owner#hasOne#App\Owner"` This will create the function `owner(){$this->hasOne('App\Owner');}`. Notice that for each arg in the third section, single quotes will be wrapped around the string.

- in crud:generate, changed `--namespace` argument to `--controller-namespace` and also added a `--model-namespace` argument for model namespaces

- updated readme.md

- pagination now uses 25 as the default instead of 15

- few typos

### CODE REVIEW

**Comment by ChangePlaces** [comment]
maybe --required-fields should become --validation. this would allow each field to have validation rules specified quickly. any validation which specifies 'required' could then not have a notnull applied to it.

**Comment by ChangePlaces** [comment]
will be working on a `--relations` option later today

**Comment by sohelamin** [comment]
Hi @ChangePlaces 

I liked your proposal.
We can add --indexes & --foreigns as you suggested.

But for --required option how will you handle the validation rules ? 
eg: --fields="body#text#required_with:title|alpha_num"

except the required things everything very promising 👍

**Comment by ChangePlaces** [comment]
Ah, I didn't know you supported that. I can work that in!

**Comment by sohelamin** [comment]
Actually people are prefer to enter less inputs thats why I made it simpler 
```--fields="title#string#required, body#text#required_with:title|alpha_num"```
Maybe we need to stick with that.

**Comment by ChangePlaces** [comment]
but what if you want validation on a field that isnt required?

File: src/Commands/CrudCommand.php [hunk 1]
```diff
@@ -60,38 +65,38 @@ public function handle()
 
         $routeGroup = $this->option('route-group');
         $this->routeName = ($routeGroup) ? $routeGroup . '/' . snake_case($name, '-') : snake_case($name, '-');
+        $perPage = $this->option('pagination') ? intval($this->option('pagination')) : 15;
```
Comment (sohelamin) [inline src/Commands/CrudCommand.php]: Instead the pagination number here you can put it default option's value
eg: ```{--pagination=15 : The amount of models per page for index pages.}```

**Comment by sohelamin** [comment]
@ChangePlaces can you leave the required fields as previous & amend the PR?
Then I can merge this. We will both find out another better way to do the required stuff later.

File: src/Commands/CrudCommand.php (hunk 1 above)
Comment (ChangePlaces) [inline src/Commands/CrudCommand.php]: how?

Comment (sohelamin) [inline src/Commands/CrudCommand.php]: On the first line ```$signature``` variable

Comment (ChangePlaces) [inline src/Commands/CrudCommand.php]: if --pagination isn't sent, will ->option('pagination') still return 15 then?

Comment (sohelamin) [inline src/Commands/CrudCommand.php]: Yes.
You just need to leave ```$perPage = $this->option('pagination'); ``` there.

Comment (ChangePlaces) [inline src/Commands/CrudCommand.php]: done

**Comment by sohelamin** [comment]
Hi @ChangePlaces Your commit has a conflict on ```src/stubs/show.blade.stub``` file please resolve and make an update on the PR.
If you don't want to do that then no problem I've already made a branch with your commits.

**Comment by ChangePlaces** [comment]
i don't know know to view the conflicts? and that really should have no conflicts?

**Comment by sohelamin** [comment]
Just for confirmation, Have you tested everything working as expected?

**Comment by ChangePlaces** [comment]
no... i like to live dangerously!

just kidding! I have!

**Comment by sohelamin** [comment]
No I have a plan to make some other changes and then I'll release it as major version like 1.2.0 (not sure yet)
You can push on branch 1.1 cause, I've made master as default for users.

I'm still thinking with the required fields cause users are more prefer to make few inputs.
We could leave it as before or only have two options eg:
```--fields="title#string, body#text"```
```--validations="title#required, published_at#date"```

And Thanks again @ChangePlaces for your awesome contributions 👍

**Comment by ChangePlaces** [comment]
well as it stands, the latter option is how it is in this pr. it allows for best flexibility - the table has required fields (which may / may not be on a form) and the form can have validation defined separately for fields which may or may not be required. it's best, trust me!

**Comment by sohelamin** [comment]
Okay let think also put the example commands here.

**Comment by ChangePlaces** [comment]
I can't push to the 1.1 branch because you've merged it! any other commits i make don't get merged!

**Pull request merged**